### PR TITLE
Factorization and improvement of permutation-based HSIC p-value computation (sequential)

### PR DESCRIPTION
**Pull request by JPelamatti**
Given that I do not seem able to solve the issues with the fully parallelized computation of the permutation based HSIC p-values (see PR 1987 https://github.com/openturns/openturns/pull/1987 , which was now reverted to a draft) before the next release, I propose these much simpler modifications.

The underlying idea of this PR is to use the same set of permutations of the output vector in order to estimate the p-values for every input dimension. This results in a fair comparison between different inputs p-values, especially for smaller data set, and as an added bonus it shaves off a bit of computation time (while still not solving the computation time issue).

Cheers

### CODE REVIEW

File: lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx [hunk 1]
```diff
@@ -152,25 +152,37 @@ void HSICEstimatorImplementation::computePValuesPermutation() const
...
+    Scalar HSIC_loc;
     UnsignedInteger count = 0;
+
     for( UnsignedInteger b = 0; b < permutationSize_; ++b)
     {
-      const Sample Yp(shuffledCopy(outputSample_));
-      const SquareMatrix W(computeWeightMatrix(Yp));
-      const Scalar HSIC_loc = computeHSICIndex(xdim, Yp, covarianceList_[dim], covarianceList_[inputDimension_], W);
+	  const Sample Yp(shuffleCollection[b]);
+      const SquareMatrix W(weightMatrixCollection[b]);
```
Comment (josephmure) [inline lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx:177]:
```suggestion
      const Sample Yp(shuffleCollection[b]);
      const SquareMatrix W(weightMatrixCollection[b]);
```

**Review by josephmure — APPROVED** [review]
Interesting improvement. Thanks Julien!

File: lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx [hunk 2]
```diff
@@ -152,25 +152,37 @@ void HSICEstimatorImplementation::computePValuesPermutation() const
...
     const Scalar HSIC_obs = computeHSICIndex(xdim, outputSample_, covarianceList_[dim], covarianceList_[inputDimension_], Wobs);
-
+    Scalar HSIC_loc;
     UnsignedInteger count = 0;
+
     for( UnsignedInteger b = 0; b < permutationSize_; ++b)
     {
-      const Sample Yp(shuffledCopy(outputSample_));
-      const SquareMatrix W(computeWeightMatrix(Yp));
-      const Scalar HSIC_loc = computeHSICIndex(xdim, Yp, covarianceList_[dim], covarianceList_[inputDimension_], W);
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx:166]: you can keep variable initialization inside loop, this is no c89 :)

File: lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx [hunk 3]
```diff
@@ -152,25 +152,37 @@ void HSICEstimatorImplementation::computePValuesPermutation() const
 {
   const SquareMatrix Wobs(computeWeightMatrix(outputSample_));
   PValuesPermutation_ = Point(inputDimension_);
+  Sample ShuffledSample;
+  Collection<Sample> shuffleCollection(permutationSize_);
+  Collection<SquareMatrix> weightMatrixCollection(permutationSize_);
+
+  for( UnsignedInteger b = 0; b < permutationSize_; ++b)
+  {
+    ShuffledSample = shuffledCopy(outputSample_);
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx:161]: same here, the variable can be declared inside the loop
it should begin with with lowercase

File: lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx (hunk 2 above)
Comment (josephmure) [reply lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx:166]: Is this recommended? Is the goal to have the variables exist only within the scope of the for loop?

**Review by josephmure — APPROVED** [review]

**Review by regislebrun — APPROVED** [review]

**Pull request merged**